### ISSUE TITLE
Suddenly can't connect to servers after system update.

### ISSUE THREAD

**Issue by tegan-lamoureux** [ProtonVPN/linux-cli#208]
**Describe the bug**
Hey there. I have a ProtonVPN Plus plan. After updating my system, I can no longer use the CLI to connect to any servers (free, p2p, secure core). I can still connect from windows dual booted on the same machine though, and also from my android phone.

**To Reproduce**
1. Upgrade Void Linux system, including kernel to 5.4.50_1.
2. CLI fails to connect to any servers.

**Expected behavior**
I can connect.

**Error Messages/Program Output/Log Messages (~/.pvpn-cli/pvpn-cli.log)**
```
▶ sudo protonvpn c -f
Connecting to US-WA#16 via UDP...
Connection failed.
```

```
▶ cat .pvpn-cli/pvpn-cli.log
[...]
2020-07-07 18:55:28,764 — protonvpn-cli — DEBUG — cli:84 — ###########################
2020-07-07 18:55:28,764 — protonvpn-cli — DEBUG — cli:85 — ### NEW PROCESS STARTED ###
2020-07-07 18:55:28,764 — protonvpn-cli — DEBUG — cli:86 — ###########################
2020-07-07 18:55:28,764 — protonvpn-cli — DEBUG — cli:87 — ['/bin/protonvpn', 'c', '-f']
2020-07-07 18:55:28,764 — protonvpn-cli — DEBUG — cli:88 — USER: tegan
2020-07-07 18:55:28,764 — protonvpn-cli — DEBUG — cli:89 — CONFIG_DIR: /home/tegan/.pvpn-cli
2020-07-07 18:55:28,768 — protonvpn-cli — DEBUG — cli:92 — Arguments
{'--cc': None, '--fastest': True, '--help': False, '--p2p': False, '--random': False, '--sc': False, '--tor': False, '--version': False, '-p': None, '<servername>': None, 'c': True, 'configure': False, 'connect': False, 'd': False, 'disconnect': False, 'examples': False, 'init': False, 'r': False, 'reconnect': False, 'refresh': False, 's': False, 'status': False}
2020-07-07 18:55:28,777 — protonvpn-cli — DEBUG — fastest:145 — Starting fastest connect
2020-07-07 18:55:28,777 — protonvpn-cli — DEBUG — disconnect:307 — Initiating disconnect
2020-07-07 18:55:28,808 — protonvpn-cli — DEBUG — is_connected:195 — Checking connection Status. OpenVPN processes: 1
2020-07-07 18:55:29,380 — protonvpn-cli — DEBUG — is_connected:195 — Checking connection Status. OpenVPN processes: 0
2020-07-07 18:55:29,412 — protonvpn-cli — DEBUG — is_connected:195 — Checking connection Status. OpenVPN processes: 0
2020-07-07 18:55:29,412 — protonvpn-cli — DEBUG — manage_dns:606 — Restoring DNS
2020-07-07 18:55:29,412 — protonvpn-cli — DEBUG — manage_dns:623 — No Backupfile found
2020-07-07 18:55:29,412 — protonvpn-cli — DEBUG — manage_ipv6:686 — Restoring ip6tables
2020-07-07 18:55:29,412 — protonvpn-cli — DEBUG — manage_ipv6:701 — No Backupfile found
2020-07-07 18:55:29,412 — protonvpn-cli — DEBUG — manage_killswitch:776 — Restoring iptables
2020-07-07 18:55:29,412 — protonvpn-cli — DEBUG — manage_killswitch:785 — No Backupfile found
2020-07-07 18:55:29,412 — protonvpn-cli — DEBUG — disconnect:338 — Disconnected
2020-07-07 18:55:29,413 — protonvpn-cli — DEBUG — call_api:36 — Initiating API Call: https://api.protonvpn.ch/vpn/logicals
2020-07-07 18:55:31,247 — protonvpn-cli — DEBUG — call_api:65 — Successful json response
2020-07-07 18:55:31,285 — protonvpn-cli — DEBUG — pull_server_data:87 — SERVER_INFO_FILE written
2020-07-07 18:55:31,290 — protonvpn-cli — DEBUG — pull_server_data:94 — last_api_call updated
2020-07-07 18:55:31,291 — protonvpn-cli — DEBUG — get_servers:101 — Reading servers from file
2020-07-07 18:55:31,297 — protonvpn-cli — DEBUG — get_fastest_server:170 — Returning fastest server with pool size 4
2020-07-07 18:55:31,298 — protonvpn-cli — DEBUG — openvpn_connect:442 — Initiating OpenVPN connection
2020-07-07 18:55:31,298 — protonvpn-cli — DEBUG — openvpn_connect:443 — Connecting to US-WA#16 via UDP
2020-07-07 18:55:31,298 — protonvpn-cli — DEBUG — get_servers:101 — Reading servers from file
2020-07-07 18:55:31,312 — protonvpn-cli — DEBUG — render_j2_template:255 — Rendered /home/tegan/.pvpn-cli/connect.ovpn from openvpn_template.j2
2020-07-07 18:55:31,317 — protonvpn-cli — DEBUG — disconnect:307 — Initiating disconnect
2020-07-07 18:55:31,347 — protonvpn-cli — DEBUG — is_connected:195 — Checking connection Status. OpenVPN processes: 0
2020-07-07 18:55:31,348 — protonvpn-cli — DEBUG — manage_dns:606 — Restoring DNS
2020-07-07 18:55:31,348 — protonvpn-cli — DEBUG — manage_dns:623 — No Backupfile found
2020-07-07 18:55:31,348 — protonvpn-cli — DEBUG — manage_ipv6:686 — Restoring ip6tables
2020-07-07 18:55:31,348 — protonvpn-cli — DEBUG — manage_ipv6:701 — No Backupfile found
2020-07-07 18:55:31,348 — protonvpn-cli — DEBUG — manage_killswitch:776 — Restoring iptables
2020-07-07 18:55:31,348 — protonvpn-cli — DEBUG — manage_killswitch:785 — No Backupfile found
2020-07-07 18:55:31,348 — protonvpn-cli — DEBUG — disconnect:347 — No connection found
2020-07-07 18:55:31,348 — protonvpn-cli — DEBUG — get_ip_info:143 — Getting IP Information
2020-07-07 18:55:31,348 — protonvpn-cli — DEBUG — call_api:36 — Initiating API Call: https://api.protonvpn.ch/vpn/location
2020-07-07 18:55:32,375 — protonvpn-cli — DEBUG — call_api:65 — Successful json response
2020-07-07 18:55:32,377 — protonvpn-cli — DEBUG — openvpn_connect:474 — OpenVPN process started
2020-07-07 18:56:17,408 — protonvpn-cli — DEBUG — openvpn_connect:522 — Connection failed after 45 Seconds
```

```
▶ cat .pvpn-cli/ovpn.log 
Tue Jul  7 18:55:32 2020 OpenVPN 2.4.9 x86_64-unknown-linux-gnu [SSL (OpenSSL)] [LZO] [LZ4] [EPOLL] [PKCS11] [MH/PKTINFO] [AEAD] built on Jul  4 2020
Tue Jul  7 18:55:32 2020 library versions: LibreSSL 3.1.3, LZO 2.10
Tue Jul  7 18:55:32 2020 Outgoing Control Channel Authentication: Using 512 bit message hash 'SHA512' for HMAC authentication
Tue Jul  7 18:55:32 2020 Incoming Control Channel Authentication: Using 512 bit message hash 'SHA512' for HMAC authentication
Tue Jul  7 18:55:32 2020 TCP/UDP: Preserving recently used remote address: [AF_INET]199.187.211.107:1194
Tue Jul  7 18:55:32 2020 Socket Buffers: R=[212992->212992] S=[212992->212992]
Tue Jul  7 18:55:32 2020 UDP link local: (not bound)
Tue Jul  7 18:55:32 2020 UDP link remote: [AF_INET]199.187.211.107:1194
Tue Jul  7 18:55:32 2020 TLS: Initial packet from [AF_INET]199.187.211.107:1194, sid=a31b8c5e f9541836
Tue Jul  7 18:55:32 2020 WARNING: this configuration may cache passwords in memory -- use the auth-nocache option to prevent this
Tue Jul  7 18:56:33 2020 TLS Error: TLS key negotiation failed to occur within 60 seconds (check your network connectivity)
Tue Jul  7 18:56:33 2020 TLS Error: TLS handshake failed
Tue Jul  7 18:56:33 2020 SIGUSR1[soft,tls-error] received, process restarting
Tue Jul  7 18:56:33 2020 Restart pause, 5 second(s)
Tue Jul  7 18:56:38 2020 TCP/UDP: Preserving recently used remote address: [AF_INET]199.187.211.107:1194
Tue Jul  7 18:56:38 2020 Socket Buffers: R=[212992->212992] S=[212992->212992]
Tue Jul  7 18:56:38 2020 UDP link local: (not bound)
Tue Jul  7 18:56:38 2020 UDP link remote: [AF_INET]199.187.211.107:1194
Tue Jul  7 18:56:38 2020 TLS: Initial packet from [AF_INET]199.187.211.107:1194, sid=6995e569 d57f4c1a
```

**Desktop (please complete the following information):**
 - OS: Void Linux, 5.4.50_1 #1 SMP PREEMPT Thu Jul 2 15:07:59 UTC 2020 x86_64 GNU/Linux 
 - Python Version: 3.8.3
 - ProtonVPN-CLI Version: 2.2.4

**Additional context**
Using the same computer I can boot into windows and connect successfully. I can also connect to proton vpn on my phone both using my data plan and on my home wifi.

Things I have tried that have not resolved the issue:
- Disabling ipv6 (`sudo sysctl -w net.ipv6.conf.all.disable_ipv6=1`)
- Trying a different router and network (tethered to my cell phone vs home wifi).
- Resetting OpenVPN credentials to new values.
- Updating my system and all packages.
- Connecting with TCP instead of UDP.

**Comment by Rafficer** [comment]
I figure you tried a reboot already?

```
Tue Jul  7 18:56:33 2020 TLS Error: TLS key negotiation failed to occur within 60 seconds (check your network connectivity)
Tue Jul  7 18:56:33 2020 TLS Error: TLS handshake failed
```

I think this seems to be what makes it fail. Why that happens, I'm not sure, though...

**Comment by tegan-lamoureux** [comment]
Thanks for getting back. Yeah I've rebooted. Also tried multiple servers in US and Europe. That seems to be the suspect line but I don't really know how to dive deeper.

Void is a bit of a fringe distro but it's a rolling release same as Arch, and should have most of the same packages in the same version states.

One difference that may be relevant is that void uses LibreSSL instead of OpenSSL. Do you know if that could be an issue if OpenVPN was built against Libre?

**Comment by tegan-lamoureux** [comment]
```
OpenVPN 2.4.9 x86_64-unknown-linux-gnu [SSL (OpenSSL)] [LZO] [LZ4] [EPOLL] [PKCS11] [MH/PKTINFO] [AEAD] built on Jul  4 2020
library versions: LibreSSL 3.1.3, LZO 2.10
Originally developed by James Yonan
Copyright (C) 2002-2018 OpenVPN Inc <sales@openvpn.net>
Compile time defines: enable_async_push=no enable_comp_stub=no enable_crypto=yes enable_crypto_ofb_cfb=yes enable_debug=yes enable_def_auth=yes enable_dlopen=unknown enable_dlopen_self=unknown enable_dlopen_self_static=unknown enable_fast_install=yes enable_fragment=yes enable_iproute2=yes enable_libtool_lock=yes enable_lz4=yes enable_lzo=yes enable_management=yes enable_multihome=yes enable_pam_dlopen=no enable_pedantic=no enable_pf=yes enable_pkcs11=yes enable_plugin_auth_pam=yes enable_plugin_down_root=yes enable_plugins=yes enable_port_share=yes enable_selinux=no enable_server=yes enable_shared=yes enable_shared_with_static_runtimes=no enable_small=no enable_static=yes enable_strict=no enable_strict_options=no enable_systemd=no enable_werror=no enable_win32_dll=yes enable_x509_alt_username=no with_aix_soname=aix with_crypto_library=openssl with_gnu_ld=yes with_mem_check=no with_sysroot=no
```

**Comment by Rafficer** [comment]
I think that's your issue. There have been issues reported before about LibreSSL ([see this](https://www.reddit.com/r/ProtonVPN/comments/gwtey6/libressl_tls_cipher/)). I'm not that much into this to be of great help, tbh. It might be worth contacting ProtonVPN support, as you probably won't be able to connect with raw OpenVPN either then.

**Comment by tegan-lamoureux** [comment]
Ah yeah. That's unfortunate but likely the issue. 

I'd still consider it a bug, though? Obviously not with this project but if ProtonVPN is using an incompatible cypher with libressl (which is a supported library for OpenVPN) that's going to alienate a lot of users (many *bsd users and anyone wanting to use LibreSSL instead of OpenSSL).

LibreSSL isn't fringe, it's much more secure and running an audited and modernized codebase. I don't want to switch to OpenSSL so this means I'll likely just have to drop ProtonVPN. Which I'm really bummed about.
 
If I wanted to escalate that to someone who knows what I'd be talking about, do you know how I'd go about that? Would ProtonVPN support get it to the right devops people who work on the servers?

**Comment by Rafficer** [comment]
Yes, that's why I want you to message support. That's the primary way of reporting bugs. They'll forward this to the appropriate infrastructure team.

Feel free to link this post as well for better understanding :)

I'm going to close this issue as it's not really a bug with the CLI. I'll keep replying if there is anything, though.

One more thing. What's the output of `openvpn --show-tls` for you?

**Comment by tegan-lamoureux** [comment]
Thanks for all your help, I'll definitely get a ticket into support.

Here's the output:
```
▶ openvpn --show-tls
Available TLS Ciphers, listed in order of preference:

For TLS 1.3 and newer (--tls-ciphersuites):

ECDHE-RSA-AES256-GCM-SHA384
ECDHE-ECDSA-AES256-GCM-SHA384
ECDHE-RSA-AES256-SHA384
ECDHE-ECDSA-AES256-SHA384
ECDHE-RSA-AES256-SHA
ECDHE-ECDSA-AES256-SHA
DHE-RSA-AES256-GCM-SHA384
DHE-RSA-AES256-SHA256
DHE-RSA-AES256-SHA
ECDHE-ECDSA-CHACHA20-POLY1305
ECDHE-RSA-CHACHA20-POLY1305
DHE-RSA-CHACHA20-POLY1305
GOST2012256-GOST89-GOST89
DHE-RSA-CAMELLIA256-SHA256
DHE-RSA-CAMELLIA256-SHA
GOST2001-GOST89-GOST89
ECDHE-RSA-AES128-GCM-SHA256
ECDHE-ECDSA-AES128-GCM-SHA256
ECDHE-RSA-AES128-SHA256
ECDHE-ECDSA-AES128-SHA256
ECDHE-RSA-AES128-SHA
ECDHE-ECDSA-AES128-SHA
DHE-RSA-AES128-GCM-SHA256
DHE-RSA-AES128-SHA256
DHE-RSA-AES128-SHA
DHE-RSA-CAMELLIA128-SHA256
DHE-RSA-CAMELLIA128-SHA

For TLS 1.2 and older (--tls-cipher):

TLS-ECDHE-RSA-WITH-AES-256-GCM-SHA384
TLS-ECDHE-ECDSA-WITH-AES-256-GCM-SHA384
TLS-ECDHE-RSA-WITH-AES-256-CBC-SHA384
TLS-ECDHE-ECDSA-WITH-AES-256-CBC-SHA384
TLS-ECDHE-RSA-WITH-AES-256-CBC-SHA
TLS-ECDHE-ECDSA-WITH-AES-256-CBC-SHA
TLS-DHE-RSA-WITH-AES-256-GCM-SHA384
TLS-DHE-RSA-WITH-AES-256-CBC-SHA256
TLS-DHE-RSA-WITH-AES-256-CBC-SHA
TLS-ECDHE-ECDSA-WITH-CHACHA20-POLY1305-SHA256
TLS-ECDHE-RSA-WITH-CHACHA20-POLY1305-SHA256
TLS-DHE-RSA-WITH-CHACHA20-POLY1305-SHA256
GOST2012256-GOST89-GOST89 (No IANA name known to OpenVPN, use OpenSSL name.)
TLS-DHE-RSA-WITH-CAMELLIA-256-CBC-SHA256
TLS-DHE-RSA-WITH-CAMELLIA-256-CBC-SHA
GOST2001-GOST89-GOST89 (No IANA name known to OpenVPN, use OpenSSL name.)
TLS-ECDHE-RSA-WITH-AES-128-GCM-SHA256
TLS-ECDHE-ECDSA-WITH-AES-128-GCM-SHA256
TLS-ECDHE-RSA-WITH-AES-128-CBC-SHA256
TLS-ECDHE-ECDSA-WITH-AES-128-CBC-SHA256
TLS-ECDHE-RSA-WITH-AES-128-CBC-SHA
TLS-ECDHE-ECDSA-WITH-AES-128-CBC-SHA
TLS-DHE-RSA-WITH-AES-128-GCM-SHA256
TLS-DHE-RSA-WITH-AES-128-CBC-SHA256
TLS-DHE-RSA-WITH-AES-128-CBC-SHA
TLS-DHE-RSA-WITH-CAMELLIA-128-CBC-SHA256
TLS-DHE-RSA-WITH-CAMELLIA-128-CBC-SHA

Be aware that that whether a cipher suite in this list can actually work
depends on the specific setup of both peers. See the man page entries of
--tls-cipher and --show-tls for more details.
```

**Comment by Rafficer** [comment]
Well, it'd be really interesting to know why `TLS_AES_256_GCM_SHA384` is missing. It's the cypher ProtonVPN uses.

**Comment by tegan-lamoureux** [comment]
Ah, so this led me down an interesting path. You guys are actually using a more secure TLS 1.3 cipher (and one of the few that's explicitly recommended) that [LibreSSL hasn't fully implemented yet](https://github.com/libressl-portable/portable/issues/228). I'm sorry for what I said earlier, this isn't a bug I'm just using a tool with a slow development cycle. :smile: 

I imagine full support is being rolled our gradually. They do say it's been rolled out fully client side but I'm also not sure why TLS_AES_256_GCM_SHA384 isn't there. I'll do more digging.

**Comment by Rafficer** [comment]
Thanks for the investigation! This is great to know. Following the other thread as well :)

**Comment by tegan-lamoureux** [comment]
Of course! So for those that may follow, it looks like while LibreSSL has implemented TLS 1.3, including the cipher we need, they haven't implemented the OpenSSL TLS API, which is where the naming of the cipher comes into play to allow the server to correctly handle the handshake.

I've also read through [the release notes here](https://www.libressl.org/releases.html) for the past couple releases, and noticed on 3.2.0: 
```
Enable TLS 1.3 server side in addition to client by default.
With this change TLS 1.3 is handled entirely on the new stack
and state machine, with fallback to the legacy stack and
state machine for older versions. Note that the OpenSSL TLS 1.3
API is not yet visible/available.
```

with emphasis on **Note that the OpenSSL TLS 1.3 API is not yet visible/available.**. Which implies it's in the works. So I guess it's a waiting game at this point. Maybe I'll get my hands dirty and see if I can help with the feature.

**Comment by 4a6f656c** [comment]
@tegan-lamoureux - I believe the issue is due to the fact that ProtonVPN is requiring P-521 for TLSv1.3 key exchange (they could allow P-384 as a lower preference, which would allow TLSv1.3 to succeed). This is not a group that is enabled by default in LibreSSL and OpenVPN releases do not provide a way to configure supported groups (there is a `tls-groups` option that was added to OpenVPN a few days ago). In the interim, a workaround is to use `tls-version-max 1.2` which will force TLSv1.2 instead of attempting TLSv1.3 negotiation (and failing).

@Rafficer - `AEAD-AES256-GCM-SHA384` (aka `TLS_AES_256_GCM_SHA384`) is present but OpenVPN does not currently know about it (`openssl ciphers -v` does however).
